### DISCUSSION TITLE
Meaning of "data too large for modulus" in request verification? (1.1.1)

### DISCUSSION THREAD

**Discussion by gingerbeard-man** [openssl/openssl#23573] [Q&A] [answered]
I have two CSRs from different clients that are structurally almost identical.  One verifies ok, the other causes OpenSSL 1.1.1w to report

```
verify failure
139871836444416:error:04067084:rsa routines:rsa_ossl_public_decrypt:data too large for modulus:crypto/rsa/rsa_ossl.c:553:
139871836444416:error:0D0C5006:asn1 encoding routines:ASN1_item_verify:EVP lib:crypto/asn1/a_verify.c:170:
```
(OpenSSL 3.0.11 is less specific and only says "Certificate request self-signature verify failure")

In both cases the RSA public key is 4096 bit and the PSS signature is 512 bytes long.  So I'm not sure what the problem is.

I see that in the failing request the numeric value of the public key modulus is less than that of the signature (starting with 0xa3 vs. 0xc0).  The verifiable request has a modulus that is greater than the signature (0xe7 vs. 0xaa).  Is this what is meant by "too large"?

Failing request:

```
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                RSA Public-Key: (4096 bit)
                Modulus:
                    00:a3:28:8b:ff:11:0c:88:5e:dc:2b:44:e3:c7:c3:
...
    Signature Algorithm: rsassaPss         
         Hash Algorithm: sha256
         Mask Algorithm: mgf1 with sha256
          Salt Length: 0x20
         Trailer Field: 0xBC (default)

         c0:b7:25:df:76:8d:b4:67:58:b3:20:77:a5:d5:87:a0:b7:e6:
...
```

Successfully verified request:

```
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (4096 bit)
                Modulus:
                    00:e7:38:bb:ec:b4:38:6f:17:14:19:1c:14:51:d0:
...
    Signature Algorithm: rsassaPss
         Hash Algorithm: sha256
         Mask Algorithm: mgf1 with sha256
         Salt Length: 0x20
         Trailer Field: 0xBC (default)

         aa:ce:8b:8a:9f:41:f6:2e:fc:50:72:9e:59:3a:af:cb:8f:e4:
...
```

**Comment by davidben** [comment, marked as answer]
Signatures are indeed required to be less than the modulus. See step 1 here:
https://datatracker.ietf.org/doc/html/rfc8017#section-5.2.2

If it is out of range, it is not a valid signature.

**Reply by gingerbeard-man** [reply]
Thank you!

I wonder how they managed to create that signature.  I assume tools like OpenSSL will normally not output an invalid signature.
